roscala/src/main/scala/coop/rchain/rosette/VirtualMachine.scala

```diff
@@ -316,33 +311,72 @@ object VirtualMachine {
     }
   }
 
-  def doRtn(state: VMState): VMState = {
-    val (isError, newState) = state.ctxt.ret(state.ctxt.rslt)(state)
+  def schedule(ctxt: Ctxt): VMTransition[Unit] =
+    modify[VMState](_.update(_ >> 'strandPool)(_ :+ ctxt))
 
-    if (isError)
-      newState.set(_ >> 'vmErrorFlag)(true)
-    else if (newState.doRtnFlag)
-      newState.set(_ >> 'doNextThreadFlag)(true)
-    else
-      newState
-  }
+  /** Return current result
+    *
+    * This returns the current result to the parent ctxt.
+    * The parent ctxt possibly gets scheduled.
+    */
+  def doRtn: VMTransition[Unit] =
+    for {
+      rslt        <- inspect[VMState, Ob](_.ctxt.rslt)
+      ctxtRet     <- Ctxt.ret(rslt).embedCtxt
+      isDoRtnFlag <- inspect[VMState, Boolean](_.doRtnFlag)
 
-  def doXmit(state: VMState): VMState = {
-    val (result, newState) = state.ctxt.trgt match {
-      case ob: StdOprn => ob.dispatch(state)
-      // TODO: Add other cases
-      case _ => (Right(Ob.NIV), state)
-    }
+      (isError, optContinuation) = ctxtRet
 
-    result match {
-      case Right(ob) if ob.is(OTsysval) =>
-        // handleException(ob, instr, state.ctxt.tag)
-        newState.set(_ >> 'doNextThreadFlag)(true)
-      case Left(DeadThread)       => newState.set(_ >> 'doNextThreadFlag)(true)
-      case _ if state.xmitData._2 => newState.set(_ >> 'doNextThreadFlag)(true)
-      case _                      => newState
-    }
-  }
+      _ <- optContinuation match {
+            case Some(ctxt) =>
+              loggerStrand.info("Schedule ctxt " + ctxt.id)
+              schedule(ctxt)
+            case None => doNothing
+          }
+
+      _ <- if (isError)
+            modify[VMState](_.copy(vmErrorFlag = true))
+          else
+            doNothing
+
+      _ <- if (isDoRtnFlag) doNextThread else doNothing
+    } yield ()
+
+  /** Dispatch current target
+    *
+    * This puts the result of the dispatch into the parent ctxt.
+    * The parent ctxt (which includes the result) possibly gets scheduled.
+    *
+    * TODO: Add unwind
+    */
+  def doXmit: VMTransition[Unit] =
+    for {
+      target         <- inspect[VMState, Ob](_.ctxt.trgt)
+      dispatchResult <- target.dispatch.embedCtxt
+      next           <- inspect[VMState, Boolean](_.xmitData._2)
+
+      (result, optContinuation) = dispatchResult
+
+      _ <- optContinuation match {
+            case Some(ctxt) =>
+              loggerStrand.info("Schedule ctxt " + ctxt.id)
+              schedule(ctxt)
+            case None =>
+              pure[VMState, Unit](())
+          }
```

Comment:
This code could be extracted to a function and reused here and above, so it could be simple:
```scala
_ <- optContinuation.map(f).getOrElse(pure(()))
```

Reply:
Done